README.md
# CytoSeg

**Automated quantification of cytoskeletal networks and cellular transport from image data**

![](examples/images/figure.png)

(A) Gray-scale confocal image of two Arabidopsis hypocotyl cells after registration and background subtraction (“original + preprocessing”).
(B) Cytoskeleton image with improved signal-to-noise ratio after cropping of the largest cell and application of tubeness filter (“filtering”).
(C) Binary cytoskeleton image after application of adaptive median threshold (“thresholding”).
(D) Skeletonised cytoskeletal structures after removal of spurious fragments (“skeletonisation + cleaning”).
(E) Overlay of skeleton image and extracted cytoskeletal network with edges colour-coded by their capacity, reflecting the average filament thickness (“network extraction”).
(F) Overview of automated pipeline for extraction of actin cytoskeletal networks from image data. 

Please see [project website for details](http://mathbiol.mpimp-golm.mpg.de/CytoSeg/).

**Instructions**

Preprocessing of actin and Golgi image data:

- Download the Fiji source-code above (fiij.py).
- Open and run the script using Fiji.
- Select the directory containing both actin and Golgi image data and adjust the file names in the script.
- Select the cellular region of interest when asked and confirm selection to proceed.
- The preprocessed images and tracking data are written to the selected directory.

Extraction of actin networks from image data:

- Download the Python source-code above (extraction.py and utils.py).
- Open the extraction script and adjust parameters as needed.
- Run the extraction script using Python (2 or 3).
- Exemplary plots and the computed network properties are written to the selected directory.

**Citation**

If this software is useful for your work, please include the following reference in your publication or redistribution:

Breuer et al. (2017, accepted) System-wide organization of actin cytoskeleton determines organelle transport in plant hypocotyl cells. 




CHANGELOG.md
# Change log

## 0.1 (2017-04-08)

  - first release


tests/test.py
################################################################################
# Module: test.py
# Description: Test imports and network extraction
# License: GPL3, see full license in LICENSE.txt
# Web: https://github.com/DavidBreuer/CytoSeg
################################################################################

#%%############################################################################# test imports

def test_imports():
    
    import itertools
    import matplotlib.pyplot as plt
    import networkx as nx
    import numpy as np
    import os
    import pandas as pd
    import random
    import scipy as sp
    import scipy.misc
    import scipy.ndimage
    import scipy.optimize
    import scipy.spatial
    import scipy.stats
    import scipy.cluster
    import skimage
    import skimage.filters
    import skimage.io
    import skimage.morphology
    import skimage.feature
    import skimage.segmentation
    import shapely
    import shapely.geometry
    import sys
    import xml
    import xml.dom
    import xml.dom.minidom   
    
    return None

#%%############################################################################# test read tiff
     
def test_read_tiff():
    
    import skimage
    import skimage.io
    im=skimage.io.imread('../examples/data/',plugin='tifffile')  
    
    return None
 
#%%############################################################################# under construction
    
    
   
    





CytoSeg/log.py
################################################################################
# Module: log.py
# Description: Logging functions
# License: GPL3, see full license in LICENSE.txt
# Web: https://github.com/DavidBreuer/CytoSeg
################################################################################

import os
import sys
import time
import unicodedata
import logging as lg
import datetime as dt
import networkx as nx
import numpy as np
import pandas as pd
import requests

def log(message, level=None, name=None, filename=None):
    """
    Write a message to the log file and/or print to the the console.

    Parameters
    ----------
    message : string
        the content of the message to log
    level : int
        one of the logger.level constants
    name : string
        name of the logger
    filename : string
        name of the log file

    Returns
    -------
    None
    """

    if level is None:
        level = globals.log_level
    if name is None:
        name = globals.log_name
    if filename is None:
        filename = globals.log_filename

    # if logging to file is turned on
    if globals.log_file:
        # get the current logger (or create a new one, if none), then log message at requested level
        logger = get_logger(level=level, name=name, filename=filename)
        if level == lg.DEBUG:
            logger.debug(message)
        elif level == lg.INFO:
            logger.info(message)
        elif level == lg.WARNING:
            logger.warning(message)
        elif level == lg.ERROR:
            logger.error(message)

    # if logging to console is turned on, convert message to ascii and print to the console
    if globals.log_console:
        # capture current stdout, then switch it to the console, print the message, then switch back to what had been the stdout
        # this prevents logging to notebook - instead, it goes to console
        standard_out = sys.stdout
        sys.stdout = sys.__stdout__

        # convert message to ascii for console display so it doesn't break windows terminals
        message = unicodedata.normalize('NFKD', make_str(message)).encode('ascii', errors='replace').decode()
        print(message)
        sys.stdout = standard_out


def get_logger(level=None, name=None, filename=None):
    """
    Create a logger or return the current one if already instantiated.

    Parameters
    ----------
    level : int
        one of the logger.level constants
    name : string
        name of the logger
    filename : string
        name of the log file

    Returns
    -------
    logger.logger
    """

    if level is None:
        level = globals.log_level
    if name is None:
        name = globals.log_name
    if filename is None:
        filename = globals.log_filename

    logger = lg.getLogger(name)

    # if a logger with this name is not already set up
    if not getattr(logger, 'handler_set', None):

        # get today's date and construct a log filename
        todays_date = dt.datetime.today().strftime('%Y_%m_%d')
        log_filename = '{}/{}_{}.log'.format(globals.logs_folder, filename, todays_date)

        # if the logs folder does not already exist, create it
        if not os.path.exists(globals.logs_folder):
            os.makedirs(globals.logs_folder)

        # create file handler and log formatter and set them up
        handler = lg.FileHandler(log_filename, encoding='utf-8')
        formatter = lg.Formatter('%(asctime)s %(levelname)s %(name)s %(message)s')
        handler.setFormatter(formatter)
        logger.addHandler(handler)
        logger.setLevel(level)
        logger.handler_set = True

    return logger


def make_str(value):
    """
    Convert a passed-in value to unicode if Python 2, or string if Python 3.

    Parameters
    ----------
    value : any
        the value to convert to unicode/string

    Returns
    -------
    unicode or string
    """
    try:
        # for python 2.x compatibility, use unicode
        return unicode(value)
    except:
        # python 3.x has no unicode type, so if error, use str type
        return str(value)

 
 


examples/example.py
################################################################################
# Module: example.py
# Description: Test imports and network extraction
# License: GPL3, see full license in LICENSE.txt
# Web: https://github.com/DavidBreuer/CytoSeg
################################################################################

#%%############################################################################# imports

import matplotlib as mpl
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import os
import pandas as pd
import random
import scipy as sp
import skimage
import skimage.io
import skimage.filters
import sys

import utils
    
#%%############################################################################# parameters

sigma=2.0                                                                       # tubeness filter width
block=101.0                                                                     # adaptive median filter block size
small=25.0                                                                      # smallest components size threshold
factr=0.5                                                                       # fraction of average intensity threshold
depth=7.75                                                                      # spacing between z-slices in xy-pixels spacings (1mum / 0.129mum/pixel = 7.75 pixels)

path='data/'                                                                    # directory with actin and Golgi images                                                        
aa='actin_filter.tif'                                                           # name of actin image 
gg='golgi_filter.tif'                                                           # name of Golgi image

#%%############################################################################# extract and randomize networks

imO=skimage.io.imread(path+aa,plugin='tifffile')[0]                             # open first frame of actin image
imT=skimage.io.imread(path+gg,plugin='tifffile')[0]                             # open Golgi image
mask=skimage.io.imread(path+'mask.tif',plugin='tifffile')>0                     # open mask

track=utils.xmlread(path+'track.xml')                                           # read Golgi tracking results
T=len(track)                                                                    # get number of tracks

print('extract','segment')  
imI=utils.im2d3d(imO)                                                           # if 2D image convert to 3D
imG=skimage.filters.gaussian(imI,sigma)                                         # apply Gaussian filter                                    
imR,imA=utils.skeletonize_graph(imG,mask,sigma,block,small,factr)               # filter and skeletonize actin image
imE=utils.node_graph(imA>0,imG)                                                 # detect filaments and network nodes
     
print('extract','graph')     
gBo,pos=utils.make_graph(imE,imG)                                               # construct graph from filament and node image
gBu=utils.unify_graph(gBo)                                                      # project multigraph to simple graph
gBc=utils.connect_graph(gBu,pos,imG)                                            # connect disconnected components of graph
gBx=utils.centralize_graph(gBc)                                                 # compute edge centrality measures
gBn=utils.normalize_graph(gBx)                                                  # normalize total edge capacity to one    
         
#%%############################################################################# plot data

print('export','plot')  

aspect=2.0                                                                      # set aspect ratio                
alpha=1.0                                                                       # set transparency
lw=1.5                                                                          # set line width        

plt.clf()
gs=mpl.gridspec.GridSpec(1,2,width_ratios=[1,1],height_ratios=[1],left=0.01,bottom=0.01,right=0.99,top=0.99,wspace=0.1,hspace=0.1)                                                       
wh=np.array(np.where(mask))[::-1]
axis=np.hstack(zip(np.nanmin(wh,1),np.nanmax(wh,1)))

plt.subplot(gs[0])                                                              # plot actin image and extracted biological network
plt.title('biological\nactin network')
plt.imshow(imO,cmap='Greys',interpolation='nearest',aspect=aspect)              
ec=1.0*np.array([d['capa'] for u,v,d in gB.edges(data=True)])
nx.draw_networkx_edges(gB,pB[:,:2],edge_color=plt.cm.jet(ec/ec.max()),width=lw,alpha=alpha)
plt.axis(axis)
plt.axis('off')

plt.subplot(gs[1])                                                              # plot Golgi image and tracks
plt.title('Golgi tracks')
plt.imshow(imT,cmap='Greys',interpolation='nearest',aspect=aspect)
for ti,t in enumerate(track[::-1]):
    plt.plot(t[:,1],t[:,2],color=plt.cm.jet(1.0*ti/T),lw=lw,alpha=0.5)
plt.axis(axis)
plt.axis('off')
 
plt.savefig(path+'out_plot.pdf')



CytoSeg/extraction.py
################################################################################
# Module: test.py
# Description: Test imports and network extraction
# License: GPL3, see full license in LICENSE.txt
# Web: https://github.com/DavidBreuer/CytoSeg
################################################################################

#%%############################################################################# imports

import matplotlib as mpl
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import os
import pandas as pd
import random
import scipy as sp
import skimage
import skimage.io
import skimage.filters
import sys

import utils
    
#%%############################################################################# parameters

sigma=2.0                                                                       # tubeness filter width
block=101.0                                                                     # adaptive median filter block size
small=25.0                                                                      # smallest component size threshold
factr=0.5                                                                       # fraction of average intensity threshold

randw=0                                                                         # randomization method (0 = shuffle nodes and edges / 1 = shuffle edge weights only)
randn=20                                                                        # number of randomized networks

depth=7.75                                                                      # spacing between z-slices in xy-pixels spacings (1mum / 0.129mum/pixel = 7.75 pixels)

path=''                                                                         # directory with actin and Golgi images                                                        
aa='actin_filter.tif'                                                           # name of actin image 
gg='golgi_filter.tif'                                                           # name of Golgi image

#%%############################################################################# extract and randomize networks

imO=skimage.io.imread(path+aa,plugin='tifffile')                                # open actin image
I=len(imO)                                                                      # get number of frames

Z=1                                                                             # set number of z-slices    
shape=imO.shape
if(len(shape)>3): 
    Z=shape[3]

imT=skimage.io.imread(path+gg,plugin='tifffile')                                # open Golgi image

track=utils.xmlread(path+'track.xml')                                           # read Golgi tracking results
T=len(track)                                                                    # get number of tracks
     
mask=skimage.io.imread(path+'mask.tif',plugin='tifffile')>0                     # open mask

R=randn                                                                         # rename number of randomized networks

#%%#

dataB=[]                                                                        # empty list for extracted networks and computed properties
dataP=[]
dataR=[]

for i in range(I):                                                          	# for each frame...

    print('extract',i,I,'segment')  
    imI=imO[i]                                                                  # get actin  image
    imI=utils.im2d3d(imI)                                                       # if 2D image convert to 3D
    imG=skimage.filters.gaussian(imI,sigma)                                     # apply Gaussian filter                                    
    imR,imA=utils.skeletonize_graph(imG,mask,sigma,block,small,factr)           # filter and skeletonize actin image
    imE=utils.node_graph(imA>0,imG)                                             # detect filaments and network nodes
             
    print('extract',i,I,'graph')     
    gBo,pos=utils.make_graph(imE,imG)                                           # construct graph from filament and node image
    gBu=utils.unify_graph(gBo)                                                  # project multigraph to simple graph
    gBc=utils.connect_graph(gBu,pos,imG)                                        # connect disconnected components of graph
    gBx=utils.centralize_graph(gBc)                                             # compute edge centrality measures
    gBn=utils.normalize_graph(gBx)                                              # normalize total edge capacity to one    
    quant=utils.compute_graph(gBn,pos,mask)                                     # compute graph properties
    dataB.append([i,gBn,pos,quant])                                             # append data
        
    for r in range(R):        

        print('extract',i,I,'randomize',r,R)  
        gRo,poz=utils.randomize_graph(gBu,pos,mask,planar=1,weights=randw)      # randomize biological network
        gRu=utils.unify_graph(gRo)                                              # project multigraph to simple graph
        gRc=utils.connect_graph(gRu,poz,imG)                                    # connect disconnected components of graph
        gRx=utils.centralize_graph(gRc)                                         # compute edge centrality measures
        gRn=utils.normalize_graph(gRx)                                          # normalize total edge capacity to one        
        quant=utils.compute_graph(gRn,poz,mask)                                 # compute graph properties         
        dataR.append([i,gRn,poz,quant])                                         # append data
    
#%%############################################################################# plot and export data

print('export','plot')  

i=0                                                                             # choose time point for plotting
r=0                                                                             # choose randomized network for plotting
gB,pB=dataB[i*1+0][1],dataB[i*1+0][2]                                           # get data for biological and randomized network
gR,pR=dataR[i*R+r][1],dataR[i*R+r][2]

plt.clf()
gs=mpl.gridspec.GridSpec(1,3,width_ratios=[1,1,1],height_ratios=[1],left=0.01,bottom=0.01,right=0.99,top=0.99,wspace=0.1,hspace=0.1)
aspect=2.0
alpha=1.0
lw=1.5
wh=np.array(np.where(mask))[::-1]
axis=np.hstack(zip(np.nanmin(wh,1),np.nanmax(wh,1)))

plt.subplot(gs[0])                                                              # plot actin image and extracted biological network
plt.title('biological\nactin network')
plt.imshow(imO[i],cmap='Greys',interpolation='nearest',aspect=aspect)              
ec=1.0*np.array([d['capa'] for u,v,d in gB.edges(data=True)])
nx.draw_networkx_edges(gB,pB[:,:2],edge_color=plt.cm.jet(ec/ec.max()),width=lw,alpha=alpha)
plt.axis(axis)
plt.axis('off')

plt.subplot(gs[1])                                                              # plot actin image and randomized network
plt.title('randomized\nactin network')
plt.imshow(imO[i],cmap='Greys',interpolation='nearest',aspect=aspect)         
ec=1.0*np.array([d['capa'] for u,v,d in gR.edges(data=True)])
nx.draw_networkx_edges(gR,pR[:,:2],edge_color=plt.cm.jet(ec/ec.max()),width=lw,alpha=alpha)
plt.axis(axis)
plt.axis('off')

plt.subplot(gs[2])                                                              # plot Golgi image and tracks
plt.title('Golgi tracks')
plt.imshow(imT[i],cmap='Greys',interpolation='nearest',aspect=aspect)
for ti,t in enumerate(track[::-1]):
    plt.plot(t[:,1],t[:,2],color=plt.cm.jet(1.0*ti/T),lw=lw,alpha=0.5)
plt.axis(axis)
plt.axis('off')
 
plt.savefig(path+'out_plot.pdf')

#%%#

print('export','track')  

idt=np.hstack([np.repeat(ti,len(t)) for ti,t in enumerate(track)])              # convert Golgi tracks to list
tracka=np.vstack([idt,np.vstack(track).T]).T
columns=['ID','t0','x0','y0','z0','avg.intensity0','tot.intensity0','quality0','diameter0','t1','x1','y1','z1','avg.intensity1','tot.intensity1','quality1','diameter1'] # name of recorded Golgi features

df=pd.DataFrame(tracka,columns=columns)                                         # save Golgi tracks as list
df.to_csv(path+'out_track.csv',sep=';',encoding='utf-8')

#%%#

print('export','graph')  

nx.write_gml(gBn,path+'out_graph.gml')                                          # save examplary actin network as graph
 
#%%#

print('export','data')  

quants=['time','# nodes','# edges','# connected components','avg. edge capacity','assortativity','avg. path length','CV path length','algebraic connectivity','CV edge angles','crossing number'] # list of computed network properties

quanta=np.array([np.hstack([d[0],d[-1]]) for d in dataB])                       # save properties of biological networks
df=pd.DataFrame(quanta,columns=quants)
df.to_csv(path+'out_biol.csv',sep=';',encoding='utf-8') 
 
quanta=np.array([np.hstack([d[0],d[-1]]) for d in dataR])                       # save properties of randomized networks
df=pd.DataFrame(quanta,columns=quants)
df.to_csv(path+'out_rand.csv',sep=';',encoding='utf-8') 




CytoSeg/fiji.py
################################################################################
# Module: fiji.py
# Description: Test imports and network extraction
# License: GPL3, see full license in LICENSE.txt
# Web: https://github.com/DavidBreuer/CytoSeg
################################################################################

#%%############################################################################# imports

import os
import sys
import time

import ij
import ij.IJ
import ij.process
import ij.macro
import ij.measure

import java.io.File

import fiji.plugin.trackmate.Settings as Settings
import fiji.plugin.trackmate.Model as Model
import fiji.plugin.trackmate.TrackMate as TrackMate
import fiji.plugin.trackmate.detection.DogDetectorFactory as DogDetectorFactory
import fiji.plugin.trackmate.tracking.sparselap.SparseLAPTrackerFactory as SparseLAPTrackerFactory
import fiji.plugin.trackmate.tracking.LAPUtils as LAPUtils
import fiji.plugin.trackmate.features.FeatureFilter as FeatureFilter
import fiji.plugin.trackmate.io.TmXmlWriter as TmXmlWriter
import fiji.plugin.trackmate.features.spot.SpotRadiusEstimatorFactory as SpotRadiusEstimatorFactory
import fiji.plugin.trackmate.features.spot.SpotIntensityAnalyzerFactory as SpotIntensityAnalyzerFactory
import fiji.plugin.trackmate.features.spot.SpotContrastAndSNRAnalyzerFactory as SpotContrastAndSNRAnalyzerFactory
import fiji.plugin.trackmate.features.edges.EdgeTargetAnalyzer as EdgeTargetAnalyzer
import fiji.plugin.trackmate.features.edges.EdgeVelocityAnalyzer as EdgeVelocityAnalyzer
import fiji.plugin.trackmate.features.track.TrackLocationAnalyzer as TrackLocationAnalyzer
import fiji.plugin.trackmate.features.track.TrackIndexAnalyzer as TrackIndexAnalyzer
import fiji.plugin.trackmate.features.track.TrackDurationAnalyzer as TrackDurationAnalyzer
import fiji.plugin.trackmate.features.track.TrackSpeedStatisticsAnalyzer as TrackSpeedStatisticsAnalyzer

#%%############################################################################# parameters

ij.IJ.run('Colors...', 'foreground=white background=black selection=magenta')   # set colors

overwrite_register=1                                                            # overwrite existing files (0 = no / 1 = yes)
overwrite_mask=1
overwrite_filter=1
overwrite_track=1

roll=50                                                                         # rolling ball radius in pixels

radius=4.0                                                                      # Golgi radius in pixels                                                            
quality=4.0                                                                     # minimum quality to accept detects blobs as Golgi
depth=7.75                                                                      # spacing between z-slices in xy-pixels spacings (1mum / 0.129mum/pixel = 7.75 pixels)

distL=24.0                                                                      # maximum linkage distance in pixels
distF=24.0                                                                      # maximum gap-closing distance in pixels
distG=5                                                                         # maximum frame gap number in pixels

path=ij.IJ.getDirectory("Choose a folder")                                      # select directory with actin and Golgi images  
aa='actin_original.tif'                                                         # name of actin image (must contain "original")
gg='golgi_original.tif'                                                         # name of Golgi image (must contain "original")

ij.IJ.run('Close All') 															# close open windows

#%%############################################################################# original -> register

if(os.path.isfile(path+gg.replace('original','register')) and overwrite_register==0): # skip step if output already exists
    print 'register','skip'
    
else:
    print 'register','load'
    for li,l in enumerate([aa,gg]):                                             # for actin and Golgi images...
        ij.IJ.open(path+l)                                                      # open image    
        ij.IJ.run('8-bit')                                                      # convert to 8-bit image
        #ij.IJ.run('Subtract Background...', 'rolling='+str(roll)+' stack')
        #ij.IJ.run('Bleach Correction', 'correction=[Simple Ratio] background=0')
    
    print 'register','merge'   
    ij.IJ.run('Merge Channels...', 'c1='+aa+' c2='+gg)                  		# merge actin and Golgi images
    
    print 'register','split'
    ij.IJ.selectWindow('RGB')													# select merged image
    ij.IJ.run('StackReg', 'transformation=[Rigid Body]')                        # register merged imgage
    ij.IJ.run('Split Channels')                                                 # split image channels
    
    print 'register','save'
    for li,l in enumerate([aa,gg]):                                             # for actin and Golgi images...
        if(li==0): ij.IJ.selectWindow('RGB (red)')
        if(li==1): ij.IJ.selectWindow('RGB (green)')
        ij.IJ.run('8-bit')                                                      # convert to 8-bit image
        ij.IJ.saveAs('Tiff',path+l.replace('original','register'))              # save registered image
    
    print 'register','close'
    #ij.IJ.selectWindow('Log')
    #ij.IJ.run('Close')
    ij.IJ.run('Close All')                                                      # close open windows

#%%############################################################################# register -> mask

if(os.path.isfile(path+'mask.tif') and overwrite_mask==0):                         # skip step if output already exists
    print 'mask','skip'
    
else:
    print 'mask','load'
    for li,l in enumerate([aa,gg]):                                             # for actin and Golgi images...
        ij.IJ.open(path+l.replace('original','register'))                       # open image  
        imp=ij.IJ.getImage()                                                    # rename image  
        imp.setTitle('img'+str(li))
        lx,ly,lc,lz,lt=imp.getDimensions()                                      # get image dimensions
        lz,lt=min(lz,lt),max(lz,lt)
        I=lc*lz*lt
        ij.IJ.run('8-bit')                                                      # convert to 8-bit image
        ij.IJ.run('Properties...', 'channels=1 slices=1 frames='+str(I)+' unit=pixel pixel_width=1.0000 pixel_height=1.0000 voxel_depth=1.0000') # concatenate z-slices and channels as time points (frames)
    
    print 'mask','concatenate'
    ij.IJ.run('Concatenate...', '  title=[Concatenated Stacks] image1=img0 image2=img1 image3=[-- None --]') # concatenate actin and Golgi stacks
    ij.IJ.run('Z Project...', 'projection=[Max Intensity]')                     # z-project combined stack
    
    print 'mask','select'
    ij.IJ.setTool('Freehand')  
    ij.IJ.runMacro("waitForUser ('Pause', 'Please select object of interest.')") # manually select cellular region of interest    
    
    print 'mask','clear'
    ij.IJ.run('Clear Outside')                                                  # clear outside selection
    ij.IJ.run('Fill', 'slice')                                                  # set selected region to uniform intensity
    ij.IJ.run('Measure')                                                        # measure intensity of selected region
    rt=ij.measure.ResultsTable.getResultsTable() 
    ma=rt.getValue('Mean', 0)
    ij.IJ.run('Divide...', 'value='+str(ma))                                    # normalize intensity of selection region to one
    
    print 'mask','save' 
    ij.IJ.run('8-bit')                                                          # convert to 8-bit image      
    ij.IJ.saveAs('Tiff',path+'mask.tif')                                        # save mask
    
    print 'mask','close'
    ij.IJ.selectWindow('Results')                                               # close open windows
    ij.IJ.run('Close')
    ij.IJ.run('Close All')

#%%############################################################################# register -> filter

if(os.path.isfile(path+gg.replace('original','filter')) and overwrite_filter==0): # skip step if output already exists
    print 'filter','skip'
    
else:    
    for li,l in enumerate([aa,gg]):                                             # for actin and Golgi images...  
        print 'filter','load'
        ij.IJ.open(path+l.replace('original','register'))                       # open image  
        imp=ij.IJ.getImage()                                                    # rename image
        imp.setTitle('image')
        lx,ly,lc,lz,lt=imp.getDimensions()                                      # get image dimensions
        lz,lt=min(lz,lt),max(lz,lt)
        I=lc*lz*lt

        print 'filter','filter'
        ij.IJ.run('8-bit')                                                      # convert to 8-bit image
        ij.IJ.run('Properties...', 'channels=1 slices=1 frames='+str(I)+' unit=pixel pixel_width=1.0000 pixel_height=1.0000 voxel_depth=1.0000') # concatenate z-slices and channels as time points (frames) 
        ij.IJ.run('Bleach Correction', 'correction=[Simple Ratio] background=0') # correct photobleaching        
        ij.IJ.run('Subtract Background...', 'rolling='+str(roll)+' stack')      # perform rolling ball background subtraction
        #ij.IJ.run('Despeckle', 'stack')                                        # apply median filter
        #ij.IJ.run('Enhance Contrast...', 'saturated=0.1 process_all use')      # enhance contrast

        print 'filter','crop'
        ij.IJ.open(path+'mask.tif')                                             # open mask
        imp=ij.IJ.getImage()                                                    # rename mask
        imp.setTitle('mask')
        ij.IJ.run('Calculator Plus', 'i1=image i2=mask operation=[Multiply: i2 = (i1*i2) x k1 + k2] k1=1 k2=0 create') # multiply image and mask
        ij.IJ.run('Stack to Hyperstack...', 'order=xyczt(default) channels=1 slices='+str(lz)+' frames='+str(lt)+' display=Grayscale') # convert to hyperstack with lz z-slices and lt time points (frames)

        print 'filter','close'  
        ij.IJ.saveAs('Tiff',path+l.replace('original','filter'))                # save filtered images
        ij.IJ.selectWindow('Log')                                               # close open windows
        ij.IJ.run('Close')
        ij.IJ.run('Close All')

#%%############################################################################# filter -> track     

if(os.path.isfile(path+'track.xml') and overwrite_track==0):                    # skip step if output already exists
    print 'track','skip'
    
else:
    print 'track','load'
    imp=ij.IJ.openImage(path+gg.replace('original','filter'))                   # open Golgi image  
    imp.show()
    lx,ly,lc,lz,lt=imp.getDimensions()                                          # get image dimensions
    lz,lt=min(lz,lt),max(lz,lt)
    ij.IJ.run('Properties...', 'channels=1 slices='+str(lz)+' frames='+str(lt)+' unit=pixel pixel_width=1.0000 pixel_height=1.0000 voxel_depth='+str(depth)) # set number of z-slices and frames and spacing between z-slices

    print 'track','setup'
    model=Model()                                                               # set tracking model factories
    settings=Settings()
    settings.setFrom(imp)
    settings.detectorFactory=DogDetectorFactory() 
    settings.detectorSettings['DO_SUBPIXEL_LOCALIZATION']=True                  # do subpixel localization
    settings.detectorSettings['RADIUS']=radius                                  # set blob radius
    settings.detectorSettings['TARGET_CHANNEL']=1                               # set target channel
    settings.detectorSettings['THRESHOLD']=1.0                                  # set detection threshold to one to exclude region outside mask
    settings.detectorSettings['DO_MEDIAN_FILTERING']=True                       # do median filtering
    settings.addSpotAnalyzerFactory(SpotIntensityAnalyzerFactory())
    settings.addSpotAnalyzerFactory(SpotRadiusEstimatorFactory())
    settings.addSpotAnalyzerFactory(SpotContrastAndSNRAnalyzerFactory())    
    settings.initialSpotFilterValue=quality
    settings.addSpotFilter(FeatureFilter('QUALITY',quality,True))  
    settings.trackerFactory=SparseLAPTrackerFactory()
    settings.trackerSettings=LAPUtils.getDefaultLAPSettingsMap()    
    settings.addEdgeAnalyzer(EdgeTargetAnalyzer())
    settings.addEdgeAnalyzer(EdgeVelocityAnalyzer())
    settings.addTrackAnalyzer(TrackSpeedStatisticsAnalyzer())
    settings.addTrackAnalyzer(TrackDurationAnalyzer())
    settings.addTrackAnalyzer(TrackIndexAnalyzer())
    settings.addTrackAnalyzer(TrackLocationAnalyzer())    
    settings.trackerSettings['LINKING_MAX_DISTANCE']=distL                      # set maximum linkage 
    settings.trackerSettings['GAP_CLOSING_MAX_DISTANCE']=distF                  # maximum gap-closing distance in pixels
    settings.trackerSettings['MAX_FRAME_GAP']=distG                             # maximum frame gap number in pixels

    print 'track','tracking'
    trackmate=TrackMate(model,settings)                                         # detect and track Golgi
    trackmate.process()   

    print 'track','save'
    out=java.io.File(path+'track.xml')                                          # save tracking results
    xw=TmXmlWriter(out)
    xw.appendModel(model)
    xw.appendSettings(settings)
    xw.writeToFile()
    ij.IJ.run('Close All')                                                      # close open windows


